reps/2023-03-10-batch-inference.md

```diff
@@ -0,0 +1,150 @@
+    feature_columns: List of columns in the preprocessed dataset to use for
+      prediction. Columns not specified will be dropped
+      from `data` before being passed to the predictor.
+      If None, use all columns in the preprocessed dataset.
+    keep_columns: List of columns in the preprocessed dataset to include
+        in the prediction result. This is useful for calculating final
+        accuracies/metrics on the result dataset. If None,
+        the columns in the output dataset will contain
+        just the prediction results.
+    min_scoring_workers: Minimum number of scoring actors.
+    max_scoring_workers: If set, specify the maximum number of scoring actors.
+    num_cpus_per_worker: Number of CPUs to allocate per scoring worker.
+      Set to 1 by default.
+    num_gpus_per_worker: Number of GPUs to allocate per scoring worker.
+      Set to 0 by default. If you want to use GPUs for inference, please
+      specify this parameter.
+    ray_remote_args: Additional resource requirements to request from ray.
+    predict_kwargs: Keyword arguments passed to the predictor's
+        ``predict()`` method.
+  """
+```
+
+3. Deprecate `Chain` Preprocessor, as chains can be represented via multiple `map_preprocessor` calls.
+4. Deprecate `BatchMapper`, as the same functionality can be achieved via `map_batches`.
+      
+## Example: Multi-stage Inference Pipeline from pre-trained model
+
+Note that Checkpoints are not on the critical path, but can still be used.
+
+### Before
+```python
+preprocessor1 = BatchMapper(lambda x: x+1)
+preprocessor2 = TorchVisionPreprocessor(torchvision.transforms.Crop(224, 224))
+chained = Chain([preprocessor1, preprocessor2])
+
+model = pretrained_resnet()
+checkpoint = TorchCheckpoint.from_model(model, preprocessor=chained)
+batch_predictor = BatchPredictor.from_checkpoint(checkpoint)
+
+ds = ds.read_parquet("s3://data")
+predictions = batch_predictor.predict(ds)
+```
+
+### After
+```python
+model = pretrained_resnet()
+predictor = TorchPredictor(model)
+
+# Alternatively, delayed model creation
+# predictor = lambda: TorchPredictor(pretrained_resnet())
+
+# Alternatively, create from AIR Checkpoint.
+# predictor = lambda: TorchPredictor.from_checkpoint(air_checkpoint)
+
+predictions = ds.read_parquet("s3://data")
+    .map_batches(lambda x: x+1)
+    .map_preprocessor(TorchVisionPreprocessor(torchvision.transforms.Crop(224, 224)))
+    .map_predict(predictor)
+```
+
+## Compatibility, Deprecation, and Migration Plan
+An important part of the proposal is to explicitly point out any compability implications of the proposed change. If there is any, we should thouroughly discuss a plan to deprecate existing APIs and migration to the new one(s).
+
+Ray 2.4: Onboard new users onto new APIs
+- Introduce `map_preprocessor` and `map_predictor` as Alpha APIs.
+- Rewrite all batch inference examples to use these APIs, and use `map_batches` directly instead of `BatchMapper` and `Chain`.
```

Comment:
This breaks training->serving compatibility, so I think we should punt.

Reply:
This is just changing the batch inference examples to the new API. The existing training->Ray Serve workflow remains the same?

Reply:
Got it. Agree batch inference specific workloads can avoid the Preprocessor class entirely.